source/common/upstream/health_discovery_service.cc

```diff
@@ -102,10 +102,29 @@ void HdsDelegate::handleFailure() {
 // TODO(lilika): Add support for the same endpoint in different clusters/ports
 envoy::service::health::v3::HealthCheckRequestOrEndpointHealthResponse HdsDelegate::sendResponse() {
   envoy::service::health::v3::HealthCheckRequestOrEndpointHealthResponse response;
+
   for (const auto& cluster : hds_clusters_) {
+    // Add cluster health response and set name
+    auto* cluster_health =
+        response.mutable_endpoint_health_response()->add_cluster_endpoints_health();
+    cluster_health->set_cluster_name(cluster->info()->name());
+
+    // Hold all endpoint health information by locality, updated when a new locality
+    // has been seen.
+    envoy::service::health::v3::LocalityEndpointsHealth* locality_health = nullptr;
+
+    // Iterate through all hosts in our priority set.
     for (const auto& hosts : cluster->prioritySet().hostSetsPerPriority()) {
       for (const auto& host : hosts->hosts()) {
-        auto* endpoint = response.mutable_endpoint_health_response()->add_endpoints_health();
+        // If we are on a new locality grouping, add grouping and copy the locality.
+        if (locality_health == nullptr ||
```

Comment:
Better to have more detailed comment for better understanding. E.g. endpoints of the same locality are placed together when created in processMessage().

Reply:
Sorry for the confusion.
I just feel it's necessary to mention the endpoints of the same locality were placed together in processMessage(), which is important for understanding the code here.
It's not necessary to explain too much logic, which is obvious from the code.